P2P: Block downloader

Boog900 opened the pull request:
This adds the block downloader, which is way more complicated than its name suggests.

The block downloader's job is to find a higher chain to follow and follow it, requesting blocks from peers and guaranteeing the blocks are yielded in order.

More details are on the docs for the `BlockDownloader`

Boog900 commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1140 @@
+    }
+}
+
+/// # Block Downloader
+///
+/// This is the block downloader, which finds a chain to follow and attempts to follow it, adding the
+/// downloaded blocks to an [`async_buffer`].
+///
+/// ## Implementation Details
+///
+/// The first step to downloading blocks is to find a chain to follow, this is done by [`initial_chain_search`],
+/// docs can be found on that function for details on how this is done.
+///
+/// With an initial list of block IDs to follow the block downloader will then look for available peers
+/// to download blocks from.
+///
+/// For each peer we will then allocate a batch of blocks for them to retrieve, as these blocks come in
+/// we add them to queue for pushing into the [`async_buffer`], once we have the oldest block downloaded
+/// we send it into the buffer, repeating this until the oldest current block is still being downloaded.
+///
+/// When a peer has finished downloading blocks we add it to our list of ready peers, so it can be used to
+/// request more data from.
+///
+/// Ready peers will either:
+/// - download the next batch of blocks
+/// - request the next chain entry
+/// - download an already requested batch of blocks this might happen due to an error in the previous request
+/// or because the queue of ready blocks is too large, so we need the oldest block to clear it.
```
here are some more details on the implementation

Boog900 commented:
actually I'm going to try add a test first

Boog900 commented:
Managed to write a test, was not fun :)

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+    async fn check_pending_peers(&mut self, chain_tracker: &mut ChainTracker<N>) {
+        tracing::debug!("Checking if we can give any work to pending peers.");
+
+        // HACK: The borrow checker doesn't like the following code if we don't do this.
+        let mut pending_peers = mem::take(&mut self.pending_peers);
```
Usually this indicates some `struct` or `fn` needs to be factored but maybe not possible here? Or at least unwieldy.

At least `BTreeMap::default()` doesn't allocate.
```suggestion
        // FIXME: We need partial mutable access to `self` (not possible) so
        // swap with an empty `BTreeMap` for the remainder of the function.
        // <https://github.com/rust-lang/rfcs/issues/1215>
        let mut pending_peers = mem::take(&mut self.pending_peers);
```

Boog900 replied:
Yeah thats what I had to do with the chain tracker, I thought it would be annoying to do it to `pending_peers` as well but it isn't actually too bad

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+                }
+
+                if let Some(peer) = self.try_handle_free_client(chain_tracker, peer).await {
+                    // This peer is ok however it does not have the data we currently need, this will only happen
+                    // because of it's pruning seed so just skip over all peers with this pruning seed.
```
```suggestion
                    // This peer is ok however it does not have the data we currently need, this will only happen
                    // because of its pruning seed so just skip over all peers with this pruning seed.
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+    /// This function will find the batch(es) that we are waiting on to clear our ready queue and sends another request
+    /// for them.
+    ///
+    /// Returns the [`ClientPoolDropGuard`] back if it doesn't have the batch according to it's pruning seed.
```
```suggestion
    /// Returns the [`ClientPoolDropGuard`] back if it doesn't have the batch according to its pruning seed.
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+            self.ready_batches_size
+        );
+
+        if self.inflight_requests.is_empty() {
+            panic!("We need requests inflight to be able to send the request again")
+        }
```
nit
```suggestion
        if self.inflight_requests.is_empty() {
            panic!("We need requests inflight to be able to send the request again");
        }
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+    /// The batch requested will depend on our current state, failed batches will be prioritised.
+    ///
+    /// Returns the [`ClientPoolDropGuard`] back if it doesn't have the data we currently need according
+    /// to it's pruning seed.
```
```suggestion
    /// to its pruning seed.
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+            if let Some(request) = self.inflight_requests.get_mut(&failed_request.0) {
+                // Check if this peer has the blocks according to their pruning seed.
+                if client_has_block_in_range(
+                    &client.info.pruning_seed,
+                    request.start_height,
+                    request.ids.len(),
+                ) {
+                    tracing::debug!("Using peer to request a failed batch");
+                    // They should have the blocks so send the re-request to this peer.
+                    let ids = request.ids.clone();
+                    let start_height = request.start_height;
+
+                    request.requests_sent += 1;
+
+                    self.block_download_tasks.spawn(
+                        async move {
+                            (
+                                start_height,
+                                request_batch_from_peer(client, ids, start_height).await,
+                            )
+                        }
+                        .instrument(tracing::debug_span!(
+                            "download_batch",
+                            start_height,
+                            attempt = request.requests_sent
+                        )),
+                    );
+
+                    // Remove the failure, we have just handled it.
+                    self.failed_batches.pop();
+
+                    return None;
+                }
+
+                break;
+            } else {
+                // We don't have the request in flight so remove the failure.
+                self.failed_batches.pop();
+            }
+        }
```
Can un-indent 1 level here:
```rust
let Some(request) = self.inflight_requests.get_mut(&failed_request.0) else {
    // We don't have the request in flight so remove the failure.
    self.failed_batches.pop();
    continue;
};

// rest of code
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+
+        self.block_download_tasks.spawn(
+            async move {
+                (
+                    block_entry_to_get.start_height,
+                    request_batch_from_peer(
+                        client,
+                        block_entry_to_get.ids,
+                        block_entry_to_get.start_height,
+                    )
+                    .await,
+                )
```
Maybe consider using a struct here instead of anonymous tuple + docs on field. More clear at definition + call site IMO.

Also stops clippy yelling at you with `type_complexity`.
```rust
/// The running block download tasks.
struct BlockDownloadTasks<N: NetworkZone> {
    /// The start height of the batch.
    start_height: u64,
    /// A result containing the batch or an error.
    result: Result<(ClientPoolDropGuard<N>, BlockBatch), BlockDownloadError>,
}

struct BlockDownloader<N: NetworkZone, S, C> {
    /* ... */
    block_download_tasks: JoinSet<BlockDownloadTasks>,
}
```
```suggestion
                BlockDownloadTasks {
                    start_height: block_entry_to_get.start_height,
                    result: request_batch_from_peer(
                        client,
                        block_entry_to_get.ids,
                        block_entry_to_get.start_height,
                    ).await,
                }
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+    /// or if we should request a batch of blocks.
+    ///
+    /// Returns the [`ClientPoolDropGuard`] back if it doesn't have the data we currently need according
+    /// to it's pruning seed.
```
```suggestion
    /// to its pruning seed.
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+        chain_tracker: &mut ChainTracker<N>,
+        client: ClientPoolDropGuard<N>,
+    ) -> Option<ClientPoolDropGuard<N>> {
+        // We send 2 requests, so if one of them is slow/ doesn't have the next chain we still have a backup.
```
```suggestion
        // We send 2 requests, so if one of them is slow or doesn't have the next chain, we still have a backup.
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+            .call(ChainSvcRequest::CumulativeDifficulty)
+            .await?
+        else {
+            panic!("Chain service returned ")
```
```suggestion
            panic!("Chain service returned");
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+            })
+            .await?
+        else {
+            panic!("Chain service returned ")
```
```suggestion
            panic!("Chain service returned");
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+                            "Received an invalid chain from peer: {}, exiting block downloader (it should be restarted).",
+                            entry.peer_who_told_us
+                        );
+                        entry.peer_who_told_us_handle.ban_peer(LONG_BAN)
```
nit but only kinda: https://stackoverflow.com/questions/26665471/are-semicolons-optional-in-rust.

`;` shows intent, i.e. "The returning value of this expression is ()". If `ban_peer()` eventually returned something other than `()` then this line would have different semantics, which wasn't the original intent.
```suggestion
                        entry.peer_who_told_us_handle.ban_peer(LONG_BAN);
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+                        return Err(BlockDownloadError::TimedOut);
+                    }
+
+                    self.failed_batches.push(Reverse(start_height))
```
```suggestion
                    self.failed_batches.push(Reverse(start_height));
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+                     // If we have no inflight requests, and we have had too many empty chain entries in a row assume the top has been found.
+                    if self.inflight_requests.is_empty() && self.amount_of_empty_chain_entries >= EMPTY_CHAIN_ENTRIES_BEFORE_TOP_ASSUMED {
+                        tracing::debug!("Failed to find any more chain entries, probably fround the top");
+                        return Ok(())
```
more of a nit on this one though
```suggestion
                        return Ok(());
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+                    // If we have no inflight requests, and we have had too many empty chain entries in a row assume the top has been found.
+                    if self.inflight_requests.is_empty() && self.amount_of_empty_chain_entries >= EMPTY_CHAIN_ENTRIES_BEFORE_TOP_ASSUMED {
+                        tracing::debug!("Failed to find any more chain entries, probably fround the top");
+                        return Ok(())
```
```suggestion
                        return Ok(());
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+/// Calculates the next amount of blocks to request in a batch.
+///
+/// Parameters:
+/// `previous_batch_size` is the size, in bytes, of the last batch,
+/// `previous_batch_len` is the amount of blocks in the last batch,
+/// `target_batch_size` is the target size, in bytes, of a batch.
```
```suggestion
/// - `previous_batch_size` is the size, in bytes, of the last batch
/// - `previous_batch_len` is the amount of blocks in the last batch
/// - `target_batch_size` is the target size, in bytes, of a batch
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader/chain_tracker.rs:
```diff
@@ -0,0 +1,190 @@
+    pub failures: usize,
+}
+
+pub enum ChainTrackerError {
+    NewEntryIsInvalid,
+    NewEntryDoesNotFollowChain,
+}
```
Docs?

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader/chain_tracker.rs:
```diff
@@ -0,0 +1,190 @@
+            usize::try_from(
+                pruning_seed
+                    .get_next_pruned_block(self.first_height, CRYPTONOTE_MAX_BLOCK_HEIGHT)
+                    // We check the first height is less than CRYPTONOTE_MAX_BLOCK_HEIGHT in response task.
+                    .unwrap()
```
`expect()`?
```suggestion
                    // We check the first height is less than CRYPTONOTE_MAX_BLOCK_HEIGHT in response task.
                    .unwrap()
```

hinto-janai commented on p2p/cuprate-p2p/src/constants.rs:
```diff
@@ -12,6 +12,12 @@ pub(crate) const OUTBOUND_CONNECTION_ATTEMPT_TIMEOUT: Duration = Duration::from_
 /// The durations of a short ban.
 pub(crate) const SHORT_BAN: Duration = Duration::from_secs(60 * 10);
 
+/// The durations of a medium ban.
+pub(crate) const MEDIUM_BAN: Duration = Duration::from_secs(60 * 60 * 24);
+
+/// The durations of a long ban.
+pub(crate) const LONG_BAN: Duration = Duration::from_secs(60 * 60 * 24 * 7);
```
Test upholding `LONG_BAN > MEDIUM_BAN > SHORT_BAN`?

hinto-janai commented on p2p/cuprate-p2p/src/constants.rs:
```diff
@@ -34,6 +40,35 @@ pub(crate) const MAX_TXS_IN_BROADCAST_CHANNEL: usize = 50;
+/// ref: <https://monero-book.cuprate.org/consensus_rules/transactions.html#transaction-size>
+pub(crate) const MAX_TRANSACTION_BLOB_SIZE: usize = 1_000_000;
+
+/// The maximum amount of block IDS allowed in a chain entry response.
```
```suggestion
/// The maximum amount of block IDs allowed in a chain entry response.
```

hinto-janai commented on p2p/cuprate-p2p/src/lib.rs:
```diff
@@ -156,6 +161,8 @@ pub struct NetworkInterface<N: NetworkZone> {
     make_connection_tx: mpsc::Sender<MakeConnectionRequest>,
     /// The address book service.
     address_book: BoxCloneService<AddressBookRequest<N>, AddressBookResponse<N>, tower::BoxError>,
+    /// The peers sync states service.
```
```suggestion
    /// The peer's sync states service.
```

hinto-janai reviewed (commented):
First pass, will re-review later.

Boog900 commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1181 @@
+        + 'static,
+    C::Future: Send + 'static,
+{
+    /// Creates a new [`BlockDownloader`]
+    fn new(
+        client_pool: Arc<ClientPool<N>>,
+
+        peer_sync_svc: S,
+        our_chain_svc: C,
+        buffer_appender: BufferAppender<BlockBatch>,
+
+        config: BlockDownloaderConfig,
+    ) -> Self {
```
yeah, It helps me to read it, I don't mind removing them now that this shouldn't be changing much

hinto-janai replied:
Up to you.

hinto-janai reviewed (commented):
Suggestion: some of the `fn`s and `struct`s here are very long:
- request_batch_from_peer
- initial_chain_search
- BlockDownloader::handle_download_batch_res
- BlockDownloader

This makes them hard to read / make sense of since there's so many variables in scope to keep in mind. A combination of extra `{}` scoping and/or factoring would be nice.

This also applies to files e.g. `block_downloader.rs` is 1194 lines, I would break it up since most of the stuff in there can be.

Boog900 commented on p2p/cuprate-p2p/src/block_downloader.rs:
```diff
@@ -0,0 +1,1194 @@
+    pub peer_handle: ConnectionHandle,
+}
+
+/// The block downloader config.
```
`BlockDownloader` isn't reachable from outside the crate

hinto-janai replied:
For IDE jump-to-definition, also shows that it's a specific struct.

Boog900 commented on p2p/cuprate-p2p/src/block_downloader/chain_tracker.rs:
```diff
@@ -0,0 +1,193 @@
+        self.first_height + u64::try_from(top_block_idx).unwrap()
+    }
+
+    /// Returns the total number of queued batches for a certain `batch_size`.
+    pub fn block_requests_queued(&self, batch_size: usize) -> usize {
```
This isn't public API so I don't think `NonZeroUsize` would be useful

Boog900 commented on p2p/monero-p2p/src/client.rs:
```diff
@@ -158,11 +159,68 @@ impl<Z: NetworkZone> Service<PeerRequest> for Client<Z> {
+                TrySendError::Closed(req) | TrySendError::Full(req) => {
+                    self.set_err(PeerError::ClientChannelClosed);
+
+                    let _ = req
+                        .response_channel
+                        .send(Err(PeerError::ClientChannelClosed.into()));
```
There is on `InfallibleOneshotReceiver`, I added a comment here as well

Boog900 commented on p2p/cuprate-p2p/src/block_downloader/chain_tracker.rs:
```diff
@@ -0,0 +1,193 @@
+    /// Returns `true` if the peer is expected to have the next block after our highest seen block
+    /// according to their pruning seed.
+    pub fn should_ask_for_next_chain_entry(&self, seed: &PruningSeed) -> bool {
+        seed.has_full_block(self.top_height(), CRYPTONOTE_MAX_BLOCK_HEIGHT)
```
I did debate it but there are only a few constants so I don't really want to build a crate just for them.

hinto-janai replied:
Someday: https://github.com/Cuprate/cuprate/issues/189.

Boog900 commented:
@hinto-janai I split up the block downloader file into separate files and simplified some of the functions, however some are still pretty long, I couldn't find an easy way to make the simpler.

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader/block_queue.rs:
```diff
@@ -0,0 +1,172 @@
+
+    /// Adds an incoming batch to the queue and checks if we can push any batches into the [`async_buffer`].
+    ///
+    /// `oldest_in_flight_start_height` Should be the start height of the oldest batch that is still inflight, if
```
```suggestion
    /// `oldest_in_flight_start_height` should be the start height of the oldest batch that is still inflight, if
```

hinto-janai commented on p2p/cuprate-p2p/src/block_downloader/chain_tracker.rs:
```diff
@@ -97,6 +107,9 @@ impl<N: NetworkZone> ChainTracker<N> {
     /// Returns the total number of queued batches for a certain `batch_size`.
+    ///
+    /// # Panics
+    /// This functions panics if `batch_size` is `0`.
```
oops, my bad.
```suggestion
    /// This function panics if `batch_size` is `0`.
```
